notation.go

```diff
@@ -90,18 +90,22 @@ func Sign(ctx context.Context, signer Signer, repo registry.Repository, signOpts
 
 	logger := log.GetLogger(ctx)
 	artifactRef := signOpts.ArtifactReference
-	if ref, err := orasRegistry.ParseReference(artifactRef); err == nil {
+	ref, err := orasRegistry.ParseReference(artifactRef)
+	if err == nil {
 		// artifactRef is a valid full reference
 		artifactRef = ref.Reference
 	}
+
 	targetDesc, err := repo.Resolve(ctx, artifactRef)
 	if err != nil {
 		return ocispec.Descriptor{}, fmt.Errorf("failed to resolve reference: %w", err)
 	}
-	if artifactRef != targetDesc.Digest.String() {
+	if ref.ValidateReferenceAsDigest() != nil {
 		// artifactRef is not a digest reference
 		logger.Warnf("Always sign the artifact using digest(`@sha256:...`) rather than a tag(`:%s`) because tags are mutable and a tag reference can point to a different artifact than the one signed", artifactRef)
 		logger.Infof("Resolved artifact tag `%s` to digest `%s` before signing", artifactRef, targetDesc.Digest.String())
+	} else if ref.Reference != targetDesc.Digest.String() {
+		return ocispec.Descriptor{}, fmt.Errorf("user input digest %s does not match the resolved digest %s", ref.Reference, targetDesc.Digest.String())
```

Comment:
Proposing an alternative solution:
```
import "github.com/opencontainers/go-digest"
```
and under `Sign`, have something like this:
```
// at this point, ref can be empty and should not be used
// artifactRef is a tag or a digest, if it's a digest it has to match
// the resolved digest.
dgst, err := digest.Parse(artifactRef)
if err != nil {
    // artifactRef is not a digest reference
    logger.Warnf("Always sign the artifact using digest(`@sha256:...`) rather than a tag(`:%s`) because tags are mutable and a tag reference can point to a different artifact than the one signed", artifactRef)
    logger.Infof("Resolved artifact tag `%s` to digest `%s` before signing", artifactRef, targetDesc.Digest.String())
} else if dgst.String() != targetDesc.Digest.String() {
    return ocispec.Descriptor{}, fmt.Errorf("user input digest %s does not match the resolved digest %s", ref.Reference, targetDesc.Digest.String())
}
```